File: frontend/authoring/performanceTelemetry.ts
const HISTOGRAM_BOUNDS = [4, 16, 50, 100, 200, 500, 1000, 5000];
export const AUTHORING_TELEMETRY_FLUSH_EVENT =
  "weblog:authoring-telemetry-flush";

type Metric = {
  name: string;
  unit: string;
  count?: number;
  sum?: number;
  min?: number;
  max?: number;
  explicit_bounds?: number[];
  bucket_counts?: number[];
  value?: number;
};

type TelemetrySpan = {
  name: "authoring.session.restore";
  duration_ms: number;
  attributes: Record<string, string | boolean>;
};

export type PerformanceTelemetryBatch = {
  schema_version: "1.0";
  resource: { attributes: Record<string, string | number | boolean> };
  metrics: Metric[];
  spans: TelemetrySpan[];
};

type StartOptions = {
  body: string;
  csrfToken: string;
  environment: string;
  serviceVersion: string;
};

type FlushEvent = CustomEvent<{ body?: string; saveDurationMs?: number }>;

type MemoryPerformance = Performance & {
  memory?: { usedJSHeapSize: number };
};

type DiscardableDocument = Document & { wasDiscarded?: boolean };

export function valueBucket(value: number): string {
  if (value === 0) return "0";
  if (value <= 10) return "1-10";
  if (value <= 100) return "11-100";
  if (value <= 1_000) return "101-1000";
  if (value <= 10_000) return "1001-10000";
  return "10001+";
}

export function histogramMetric(name: string, values: number[]): Metric | null {
  if (values.length === 0) return null;
  const bucketCounts = Array.from(
    { length: HISTOGRAM_BOUNDS.length + 1 },
    () => 0,
  );
  for (const value of values) {
    const index = HISTOGRAM_BOUNDS.findIndex((bound) => value <= bound);
    bucketCounts[index === -1 ? HISTOGRAM_BOUNDS.length : index] += 1;
  }
  return {
    name,
    unit: "ms",
    count: values.length,
    sum: values.reduce((total, value) => total + value, 0),
    min: Math.min(...values),
    max: Math.max(...values),
    explicit_bounds: HISTOGRAM_BOUNDS,
    bucket_counts: bucketCounts,
  };
}

export function startAuthoringPerformanceTelemetry(
  options: StartOptions,
): () => void {
  const interactions = new Map<number, number>();
  const durations = new Map<string, number[]>();
  const requestCounts = new Map<string, number>();
  const spans: TelemetrySpan[] = [];
  const observers: PerformanceObserver[] = [];
  let latestBody = options.body;
  let stopped = false;
  const wasDiscarded = Boolean((document as DiscardableDocument).wasDiscarded);

  const recordDuration = (name: string, duration: number) => {
    const values = durations.get(name) || [];
    values.push(duration);
    durations.set(name, values);
  };

  const observe = (
    type: string,
    callback: (entry: PerformanceEntry) => void,
  ) => {
    if (!PerformanceObserver.supportedEntryTypes.includes(type)) return;
    const observer = new PerformanceObserver((list) =>
      list.getEntries().forEach(callback),
    );
    observer.observe({ type, buffered: true });
    observers.push(observer);
  };

  observe("event", (entry) => {
    const interactionId =
      "interactionId" in entry ? Number(entry.interactionId) : 0;
    if (interactionId > 0)
      interactions.set(
        interactionId,
        Math.max(interactions.get(interactionId) || 0, entry.duration),
      );
  });
  observe("longtask", (entry) =>
    recordDuration("authoring.long_task.duration", entry.duration),
  );
  observe("resource", (entry) => {
    if (!entry.name.includes("/api/")) return;
    if (entry.name.includes("/api/embed")) {
      recordDuration("authoring.requests.embed.duration", entry.duration);
      requestCounts.set(
        "authoring.requests.embed.total",
        (requestCounts.get("authoring.requests.embed.total") || 0) + 1,
      );
    }
  });

  if (wasDiscarded) {
    spans.push({
      name: "authoring.session.restore",
      duration_ms: performance.now(),
      attributes: {
        "session.discarded": true,
        "navigation.type":
          (
            performance.getEntriesByType("navigation")[0] as
              | PerformanceNavigationTiming
              | undefined
          )?.type || "unknown",
      },
    });
  }

  const flush = async () => {
    if (stopped) return;
    try {
      const storageStartedAt = performance.now();
      sessionStorage.setItem("weblog:telemetry-probe", "1");
      sessionStorage.removeItem("weblog:telemetry-probe");
      recordDuration(
        "authoring.session_storage.duration",
        performance.now() - storageStartedAt,
      );
    } catch (_error) {
      // Storage can be unavailable in restricted browser contexts.
    }
    const interactionValues = [...interactions.values()];
    interactions.clear();
    if (interactionValues.length > 0)
      durations.set("authoring.interaction.duration", interactionValues);

    const metrics = [...durations.entries()]
      .map(([name, values]) => histogramMetric(name, values))
      .filter((metric): metric is Metric => metric !== null);
    durations.clear();
    requestCounts.forEach((value, name) => {
      metrics.push({ name, unit: "{request}", value });
    });
    requestCounts.clear();
    metrics.push({
      name: "authoring.dom.nodes",
      unit: "{node}",
      value: document.getElementsByTagName("*").length,
    });
    metrics.push({
      name: "authoring.frames",
      unit: "{frame}",
      value: document.querySelectorAll("iframe").length,
    });
    metrics.push({
      name: "authoring.documents",
      unit: "{document}",
      value: 1 + document.querySelectorAll("iframe").length,
    });
    const heap = (performance as MemoryPerformance).memory?.usedJSHeapSize;
    if (heap !== undefined)
      metrics.push({
        name: "authoring.js_heap.bytes",
        unit: "By",
        value: heap,
      });

    const batch: PerformanceTelemetryBatch = {
      schema_version: "1.0",
      resource: {
        attributes: {
          "service.name": "weblog-authoring",
          "service.version": options.serviceVersion,
          "deployment.environment.name": options.environment,
          "browser.time_origin_unix_ms": performance.timeOrigin,
          "session.discarded": wasDiscarded,
          "navigation.type":
            (
              performance.getEntriesByType("navigation")[0] as
                | PerformanceNavigationTiming
                | undefined
            )?.type || "unknown",
          "page.body_length_bucket": valueBucket(latestBody.length),
          "page.external_link_count_bucket": valueBucket(
            (latestBody.match(/https?:\/\//g) || []).length,
          ),
          "page.image_count_bucket": valueBucket(
            (latestBody.match(/!\[/g) || []).length,
          ),
          "page.wiki_link_count_bucket": valueBucket(
            (latestBody.match(/\[\[/g) || []).length,
          ),
        },
      },
      metrics,
      spans: spans.splice(0),
    };

    try {
      await fetch("/api/authoring/telemetry", {
        method: "POST",
        headers: {
          "content-type": "application/json",
          "x-csrf-token": options.csrfToken,
        },
        body: JSON.stringify(batch),
        keepalive: true,
      });
    } catch (_error) {
      // Telemetry must not affect authoring.
    }
  };

  const interval = window.setInterval(() => void flush(), 60_000);
  const handlePageHide = () => void flush();
  const handleBatchBoundary = (event: Event) => {
    const detail = (event as FlushEvent).detail;
    if (detail?.body !== undefined) latestBody = detail.body;
    if (detail?.saveDurationMs !== undefined) {
      recordDuration("authoring.requests.save.duration", detail.saveDurationMs);
      requestCounts.set(
        "authoring.requests.save.total",
        (requestCounts.get("authoring.requests.save.total") || 0) + 1,
      );
    }
    void flush();
  };
  window.addEventListener("pagehide", handlePageHide);
  window.addEventListener(AUTHORING_TELEMETRY_FLUSH_EVENT, handleBatchBoundary);
  if (wasDiscarded) void flush();
  return () => {
    window.clearInterval(interval);
    window.removeEventListener("pagehide", handlePageHide);
    window.removeEventListener(
      AUTHORING_TELEMETRY_FLUSH_EVENT,
      handleBatchBoundary,
    );
    observers.forEach((observer) => {
      observer.disconnect();
    });
    void flush().finally(() => {
      stopped = true;
    });
  };
}

File: lib/weblog_authoring/performance_telemetry.rb
# frozen_string_literal: true

require "base64"
require "json"
require "rack/utils"
require "time"

module WeblogAuthoring
  class PerformanceTelemetry
    MAX_BODY_BYTES = 64 * 1024
    MAX_METRICS = 24
    MAX_SPANS = 8
    ALLOWED_ATTRIBUTE_KEYS = %w[
      deployment.environment.name
      browser.time_origin_unix_ms
      navigation.type
      page.body_length_bucket
      page.external_link_count_bucket
      page.image_count_bucket
      page.wiki_link_count_bucket
      service.name
      service.version
      session.discarded
    ].freeze
    ALLOWED_METRIC_NAMES = %w[
      authoring.dom.nodes
      authoring.documents
      authoring.frames
      authoring.interaction.duration
      authoring.js_heap.bytes
      authoring.long_task.duration
      authoring.requests.embed.duration
      authoring.requests.embed.total
      authoring.requests.save.duration
      authoring.requests.save.total
      authoring.session_storage.duration
    ].freeze
    ALLOWED_SPAN_NAMES = %w[authoring.session.restore].freeze
    JSON_HEADERS = { "content-type" => "application/json; charset=utf-8" }.freeze

    def self.validate_payload!(payload)
      allocate.send(:validate_payload!, payload)
    end

    def initialize(session_codec:, allowed_github_user_id:, logger: $stdout, clock: Time.method(:now))
      @session_codec = session_codec
      @allowed_github_user_id = allowed_github_user_id
      @logger = logger
      @clock = clock
    end

    def call(event)
      return response(404, error: "Not Found") unless post_request?(event)
      return response(413, error: "Telemetry payload is too large") if decoded_body(event).bytesize > MAX_BODY_BYTES

      session = session_from(event)
      return response(401, error: "Authentication required") unless allowed_session?(session)
      return response(403, error: "CSRF token mismatch") unless secure_equal?(session.fetch("csrf_token", ""), csrf_token(event))

      payload = JSON.parse(decoded_body(event))
      validate_payload!(payload)
      restore_max_ms = payload.fetch("spans", []).filter_map do |span|
        span["duration_ms"] if span["name"] == "authoring.session.restore"
      end.max
      @logger.puts(JSON.generate(
        "event" => "authoring_performance_telemetry",
        "received_at" => @clock.call.iso8601(6),
        "restore_max_ms" => restore_max_ms,
        "payload" => payload,
      ))
      response(202, status: "accepted")
    rescue JSON::ParserError
      response(422, error: "Telemetry payload must be valid JSON")
    rescue ArgumentError => error
      response(422, error: error.message)
    end

    private

    def post_request?(event)
      event.dig("requestContext", "http", "method") == "POST" && event["rawPath"] == "/api/authoring/telemetry"
    end

    def validate_payload!(payload)
      raise ArgumentError, "Telemetry payload must be an object" unless payload.is_a?(Hash)
      raise ArgumentError, "Unsupported telemetry schema" unless payload["schema_version"] == "1.0"

      attributes = payload.fetch("resource", {}).fetch("attributes", {})
      validate_attributes!(attributes)
      metrics = payload.fetch("metrics", [])
      spans = payload.fetch("spans", [])
      raise ArgumentError, "Too many metrics" unless metrics.is_a?(Array) && metrics.length <= MAX_METRICS
      raise ArgumentError, "Too many spans" unless spans.is_a?(Array) && spans.length <= MAX_SPANS

      metrics.each { |metric| validate_metric!(metric) }
      spans.each { |span| validate_span!(span) }
    end

    def validate_attributes!(attributes)
      raise ArgumentError, "Telemetry attributes must be an object" unless attributes.is_a?(Hash)
      raise ArgumentError, "Unsupported telemetry attribute" unless (attributes.keys - ALLOWED_ATTRIBUTE_KEYS).empty?
      valid_values = attributes.values.all? do |value|
        value.is_a?(String) || value == true || value == false || (value.is_a?(Numeric) && value.finite?)
      end
      raise ArgumentError, "Telemetry attribute value is invalid" unless valid_values
      raise ArgumentError, "Telemetry attribute value is too long" if attributes.values.any? { |value| value.is_a?(String) && value.bytesize > 64 }
    end

    def validate_metric!(metric)
      raise ArgumentError, "Telemetry metric must be an object" unless metric.is_a?(Hash)
      raise ArgumentError, "Unsupported telemetry metric" unless ALLOWED_METRIC_NAMES.include?(metric["name"])
      allowed_keys = %w[name unit count sum min max explicit_bounds bucket_counts value]
      raise ArgumentError, "Unsupported telemetry metric field" unless (metric.keys - allowed_keys).empty?
      numeric_values = metric.values_at("count", "sum", "min", "max", "value").compact
      raise ArgumentError, "Telemetry metric value is invalid" unless numeric_values.all? { |value| value.is_a?(Numeric) && value.finite? }
      bounds = metric.fetch("explicit_bounds", [])
      counts = metric.fetch("bucket_counts", [])
      raise ArgumentError, "Telemetry histogram is invalid" unless bounds.is_a?(Array) && counts.is_a?(Array) && bounds.length <= 16 && counts.length <= 17
      raise ArgumentError, "Telemetry histogram value is invalid" unless (bounds + counts).all? { |value| value.is_a?(Numeric) && value.finite? && value >= 0 }
    end

    def validate_span!(span)
      raise ArgumentError, "Telemetry span must be an object" unless span.is_a?(Hash)
      raise ArgumentError, "Unsupported telemetry span" unless ALLOWED_SPAN_NAMES.include?(span["name"])
      raise ArgumentError, "Unsupported telemetry span field" unless (span.keys - %w[name duration_ms attributes]).empty?
      duration = span["duration_ms"]
      raise ArgumentError, "Telemetry span duration is invalid" unless duration.is_a?(Numeric) && duration.finite? && duration >= 0
      validate_attributes!(span.fetch("attributes", {}))
    end

    def session_from(event)
      raw = Array(event["cookies"]).find { |cookie| cookie.start_with?("weblog_authoring_session=") }
      @session_codec.read(raw.to_s.delete_prefix("weblog_authoring_session="), kind: "session")
    end

    def allowed_session?(session)
      session && Integer(session.fetch("github_user_id")) == @allowed_github_user_id
    rescue ArgumentError, KeyError, TypeError
      false
    end

    def csrf_token(event)
      event.fetch("headers", {}).fetch("x-csrf-token", "")
    end

    def decoded_body(event)
      body = event.fetch("body", "").to_s
      event["isBase64Encoded"] ? Base64.decode64(body) : body
    end

    def secure_equal?(expected, supplied)
      expected = expected.to_s
      supplied = supplied.to_s
      !expected.empty? && expected.bytesize == supplied.bytesize && Rack::Utils.secure_compare(expected, supplied)
    end

    def response(status, payload)
      { statusCode: status, headers: JSON_HEADERS, body: JSON.generate(payload) }
    end
  end
end
